### ISSUE TITLE
Projected item states are not undoable

### ISSUE THREAD

**Issue by blitzmann** [pyfa-org/Pyfa#1812]
You cannot  currently undo projected state changed. Unsrue if this was just missed back when I refactored fitting things to command pattern, or if it was skipped due to complexity and pushed off for later.

**Comment by DarkFenX** [comment]
Fixed in dca2db5a6d4d3dd44d77b3c7d12affa4928ed46b

**Comment by blitzmann** [comment]
@DarkFenX 

```def Do(self):
        fit = Fit.getInstance().getFit(self.fitID)
        if isinstance(self.thing, FitType):
            success = self.internal_history.Submit(FitToggleProjectedFitCommand(self.fitID, self.thing.ID))
        elif isinstance(self.thing, ModuleType):
            position = fit.projectedModules.index(self.thing)
            success = self.internal_history.Submit(FitToggleProjectedModuleCommand(self.fitID, position, self.click))
       ...
```

Since SQLAlchemy objects are volatile with respect to commands,  I think the  various `if isinstance(self.thing, type):` things will break once the object is deleted by another command. Lets say you toggle a projected drone, the `Do()` will work because the drone exists. However, if you then delete the drone (removes SQLAlchemy object) and undo that delete (re-creates drone as a new SQL alchemy object), that `GuiToggleProjectedCommand` is now potentially broken because `self.thing` references an object that is now deleted. That's why other commands use position or other identifiers instead of literal objects.

Now that I think about it, since you're simply checking the type and aren't doing anything to that object itself, it probably would be fine, but we'd be holding onto the dead object for the life of the command. I can't test right now unfortunately. I would maybe do the instance checking in the `__init__` and set a property like `self.type = 'Fit'` so that the command object doesn't need to carry around the volatile SQLAlchemy object

**Comment by DarkFenX** [comment]
Yes i understand that very well, that's why I am using primitives wherever i can. Here i do not store this "thing" to use it later, it's consumed instantly given pyfa creates command and "does" it right away. It won't be reused later. If these two actions would occur with extended period of time between them, I'd just run this logic in constructor and store item type as text/ID and additional data like position, on command itself. Or if there's some logic which reruns "do", we could do that as well.

Probably it'd be worth to do it this way because of safety, I just slacked it. Will change later.

**Comment by DarkFenX** [comment]
Fixed in 69ada4f3adb8638aa92a053f0c059cfacd18fa9c
